File: ruby/lib/oppex_sdk/internal/rate_limited_drop_logger.rb
# frozen_string_literal: true

module Oppex
  module Internal
    # Counts every dropped incident but reports at most one summary per interval.
    #
    # A saturated queue drops continuously, so logging each drop would replace
    # one overload with another. Not public API.
    class RateLimitedDropLogger
      INTERVAL_SECONDS = 60.0

      def initialize(interval_seconds: INTERVAL_SECONDS, clock: -> { Process.clock_gettime(Process::CLOCK_MONOTONIC) })
        @interval_seconds = interval_seconds
        @clock = clock
        @mutex = Mutex.new
        @dropped = 0
        @started_at = clock.call
      end

      # Records one drop.
      #
      # Returns the accumulated count when the interval has elapsed and the
      # caller should emit a summary, or nil otherwise. Returning the count
      # instead of logging keeps this class free of any logging decision, which
      # is also what makes it testable without capturing output.
      def record_drop
        @mutex.synchronize do
          @dropped += 1
          now = @clock.call
          next nil if now - @started_at < @interval_seconds

          dropped = @dropped
          @dropped = 0
          @started_at = now
          dropped
        end
      end
    end
  end
end

File: dotnet/src/Oppex.Integration.Sdk/Internal/RateLimitedDropLogger.cs
namespace Oppex.Integration.Sdk.Internal;

/// <summary>
/// Counts every dropped incident but reports at most one summary per interval.
/// Not public API.
/// </summary>
/// <remarks>
/// A saturated queue drops continuously, so logging each drop would replace one
/// overload with another. Returning the count instead of logging keeps this class
/// free of any logging decision, which is also what makes it testable without
/// capturing output.
/// </remarks>
internal sealed class RateLimitedDropLogger(TimeProvider timeProvider, TimeSpan? interval = null)
{
    private static readonly TimeSpan DefaultInterval = TimeSpan.FromMinutes(1);

    private readonly TimeSpan _interval = interval ?? DefaultInterval;
    private readonly Lock _gate = new();
    private long _startedAtTicks = timeProvider.GetTimestamp();
    private int _dropped;

    /// <summary>
    /// Records one drop, returning the accumulated count when the interval has
    /// elapsed and the caller should report a summary, or null otherwise.
    /// </summary>
    internal int? RecordDrop()
    {
        lock (_gate)
        {
            _dropped++;
            var now = timeProvider.GetTimestamp();
            if (timeProvider.GetElapsedTime(_startedAtTicks, now) < _interval)
            {
                return null;
            }

            var dropped = _dropped;
            _dropped = 0;
            _startedAtTicks = now;
            return dropped;
        }
    }
}
